src/pegins.cpp

```diff
@@ -0,0 +1,333 @@
+    // Must include all elements
+    if (stack.size() != 6) {
+        return false;
+    }
+
+    CDataStream stream(stack[0], SER_NETWORK, PROTOCOL_VERSION);
+    CAmount value;
+    try {
+        stream >> value;
+    } catch (...) {
+        return false;
+    }
+
+    if (!MoneyRange(value)) {
+        return false;
+    }
+
+    //TODO(rebase) CA
+    //// Get asset type
+    //if (stack[1].size() != 32) {
+    //    return false;
+    //}
+    //CAsset asset(stack[1]);
+
+    // Get genesis blockhash
+    if (stack[2].size() != 32) {
+        return false;
+    }
+    uint256 gen_hash(stack[2]);
+
+    // Get claim_script, sanity check size
+    CScript claim_script(stack[3].begin(), stack[3].end());
+    if (claim_script.size() > 100) {
+        return false;
+    }
+
+    uint256 block_hash;
+    uint256 tx_hash;
+    int num_txs;
+    // Get txout proof
+    if (Params().GetConsensus().ParentChainHasPow()) {
+        Sidechain::Bitcoin::CMerkleBlock merkle_block_pow;
+        if (!GetBlockAndTxFromMerkleBlock(block_hash, tx_hash, merkle_block_pow, stack[5])) {
+            return false;
+        }
+        if (!CheckParentProofOfWork(block_hash, merkle_block_pow.header.nBits, Params().GetConsensus())) {
+            return false;
+        }
+
+        Sidechain::Bitcoin::CTransactionRef pegtx;
+        if (!CheckPeginTx(stack[4], pegtx, prevout, value, claim_script)) {
+            return false;
+        }
+
+        num_txs = merkle_block_pow.txn.GetNumTransactions();
+    } else {
+        //TODO(rebase) parent signed blocks
+        //CMerkleBlock merkle_block;
+        //if (!GetBlockAndTxFromMerkleBlock(block_hash, tx_hash, merkle_block, stack[5])) {
+        //    return false;
+        //}
+
+        //if (!CheckProofSignedParent(merkle_block.header, Params().GetConsensus())) {
+        //    return false;
+        //}
+
+        //CTransactionRef pegtx;
+        //if (!CheckPeginTx(stack[4], pegtx, prevout, value, claim_script)) {
+        //    return false;
+        //}
+
+        //num_txs = merkle_block.txn.GetNumTransactions();
+    }
+
+    // Check that the merkle proof corresponds to the txid
+    if (prevout.hash != tx_hash) {
+        return false;
+    }
+
+    // Check the genesis block corresponds to a valid peg (only one for now)
+    if (gen_hash != Params().ParentGenesisBlockHash()) {
+        return false;
+    }
+
+    //TODO(rebase) CA
+    //// Check the asset type corresponds to a valid pegged asset (only one for now)
+    //if (asset != Params().GetConsensus().pegged_asset) {
+    //    return false;
+    //}
+
+    // Finally, validate peg-in via rpc call
+    if (check_depth && gArgs.GetBoolArg("-validatepegin", DEFAULT_VALIDATE_PEGIN)) {
+        if (!IsConfirmedBitcoinBlock(block_hash, Params().GetConsensus().pegin_min_depth, num_txs)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+// Constructs unblinded output to be used in amount and scriptpubkey checks during pegin
```

Comment:
suggested rewording: Constructs unblinded "bitcoin" output to be used in amount and scriptpubkey checks during pegin validation

Reply:
done